File: cpre388/MemoryLeak/app/src/main/java/edu/iastate/cpre388/memoryleak/MainActivity.java
package edu.iastate.cpre388.memoryleak;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private final int NUM_CATS = 1000000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        BigRedButton button = new BigRedButton(this);
        button.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    for(int i = 0; i < NUM_CATS; i++) {
                        Cat cat = new Cat(getString(R.string.cat_name),
                                7,
                                getString(R.string.cat_color));
                    }
                    Toast.makeText(getApplicationContext(),
                            getString(R.string.cat_toast, NUM_CATS),
                            Toast.LENGTH_LONG).show();
                }
        });

        LinearLayout layout = findViewById(R.id.linearLayout);
        layout.addView(button);
    }
}

File: cpre388/Stopwatch/app/src/main/java/com/example/stopwatch/MainActivity.kt
package com.example.stopwatch

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider


class MainActivity : AppCompatActivity() {

    lateinit var viewModel : Timer
    var running = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        viewModel = ViewModelProvider(this).get(Timer::class.java)

        var startButton = findViewById<Button>(R.id.startButton)
        var resetButton = findViewById<Button>(R.id.resetButton)
        var timerText = findViewById<TextView>(R.id.timerText)

        viewModel.time.observe(this, Observer {
            time -> var format = formatTime(time)
            timerText.text = format
        })

        startButton.setOnClickListener {
            if(running) {
                viewModel.stopTimer()
                startButton.text = getString(R.string.start)
            }
            else {
                viewModel.startTimer()
                startButton.text = getString(R.string.stop)
            }

            // Running toggle
            running = !running
        }

        resetButton.setOnClickListener {
            viewModel.resetTimer()
            startButton.text = "Start"
            running = false
        }

        //Handler(Looper.getMainLooper()).postDelayed({

        //}, 3000)
    }

    fun formatTime(time : Long): String {
        var hours = (time / 3600000).toString().padStart(2, '0')
        var min = ((time % 3600000) / 60000).toString().padStart(2, '0')
        var sec = ((time % 60000) / 1000).toString().padStart(2, '0')
        var millis = ((time % 1000) / 100).toString()

        return "$hours:$min:$sec:$millis"
    }
}